### ISSUE TITLE
[CRASH] in db_insert_ucontact (usrloc?)

### ISSUE THREAD

**Issue by netaskd** [OpenSIPS/opensips#1640]
**OpenSIPS version you are running**

```
opensips -V
version: opensips 2.4.5 (x86_64/linux)
flags: STATS: On, DISABLE_NAGLE, USE_MCAST, SHM_MMAP, PKG_MALLOC, F_MALLOC, FAST_LOCK-ADAPTIVE_WAIT
ADAPTIVE_WAIT_LOOPS=1024, MAX_RECV_BUFFER_SIZE 262144, MAX_LISTEN 16, MAX_URI_SIZE 1024, BUF_SIZE 65535
poll method support: poll, epoll, sigio_rt, select.
main.c compiled on 01:36:45 Mar  8 2019 with gcc 4.8.5
```

**Crash Core Dump**

bt full: https://pastebin.com/x8PYn3uY
f 3, p *ptr: https://pastebin.com/wnaAh6i7

**Describe the traffic that generated the bug**

**To Reproduce**

`opensips -M1024 -m1024 -Ff /etc/opensips/opensips.cfg`
usrloc config:
```
modparam("usrloc", "nat_bflag", "uas_nat")
# FIXME: md_registrar supported only it mode, but will be fixed. Check it
modparam("usrloc", "working_mode_preset", "single-instance-sql-write-back")
modparam("usrloc", "matching_mode", 0) # 0 - contact, 1 - contact + callid
modparam("usrloc", "desc_time_order", 1)
modparam("usrloc", "cseq_delay", 5)
modparam("usrloc", "skip_replicated_db_ops", 1)
modparam("usrloc", "timer_interval", 10) # defines the periodic timer for checking the registrations status
modparam("usrloc", "use_domain", 1)
modparam("usrloc", "db_url", "mysql://to_localhost_url")
```
**Relevant System Logs**

**OS/environment information**
 - Operating System: 
`CentOS Linux release 7.6.1810 (Core) `  
`3.10.0-957.10.1.el7.x86_64`
 - OpenSIPS installation: 
`rpm`
 - other relevant information:
**Additional context**

**Comment by bogdan-iancu** [comment]
@netaskd , thanks for your report . The backtrace shows that the crash happens during OpenSIPS shutdown - could you confirm that ?
Also , in frame 0, could you please print `*_c`. The GDB commands are:
`f 0`
`p *_c`

**Comment by bogdan-iancu** [comment]
BTW, do you `registrar` or `mid_registrar` on top of `usrloc` ?

**Comment by netaskd** [comment]
Hi @bogdan-iancu 
1. f 0 p *_c https://pastebin.com/tpxcz41v
I don't do shutdown from scripts, but sometime I see running: `main -> handle_sigs -> shutdown_opensips -> cleanup -> fm_status (segfalt)` in core file, don't know what why it happens.
2. I use `registrar` module

**Comment by bogdan-iancu** [comment]
The backtrace you posted is from a crash during the opensips shutdown - the main question is why OpenSIPS doing shutdown ? IF you do not explicitly stop it, it means there is another prior crash in a different process, crash that triggers the overall OpenSIPS shutdown and eventually another crash . 
But it seems you reported the second crash (which may not be relevant in such cases) - please check for logs (opensips, syslog) and find why opensips is shutting down. Also check for other core files (maybe from the first crash?).

**Comment by netaskd** [comment]
I don't have any idea about why it shutdown. But i see that before crash shmem increases at times.
before segfalt I see:
```
shmem:total_size:: 33554432
shmem:used_size:: 11068615982480491281
shmem:real_used_size:: 11068615982482135993
shmem:max_used_size:: 18446603852886271847
shmem:free_size:: 7378128091260970055
shmem:fragments:: 66433
```
but in normal state (Up time:: 1381 [sec]) i see:
```
shmem:total_size:: 33554432
shmem:used_size:: 9283592
shmem:real_used_size:: 10934952
shmem:max_used_size:: 11328328
shmem:free_size:: 22619480
shmem:fragments:: 67038
```
crash log has no any information about additional core files and looks like:
```
Mar 26 13:02:27 [5563] CRITICAL:core:sig_usr: segfault in process pid: 5563, id: 10
Mar 26 13:02:27 [5572] CRITICAL:core:handle_tcp_worker: dead tcp worker 0 (EOF received), pid 5563
Mar 26 13:02:27 [5553] INFO:core:handle_sigs: child process 5563 exited by a signal 11
Mar 26 13:02:27 [5553] INFO:core:handle_sigs: core was generated
Mar 26 13:02:27 [5553] INFO:core:handle_sigs: terminating due to SIGCHLD
Mar 26 13:02:27 [5555] INFO:core:sig_usr: signal 15 received
```
core id 5563 looks like:
bt full: https://pastebin.com/CT778YbW

I know that it not looks like about topic start message, but maybe it will help.

**Comment by bogdan-iancu** [comment]
Thanks @netaskd , that's what I need. The original/first crash points to shm memory corruption. To catch something like this, you need to re-compile OpenSIPS with memory debugger support. For how to do it, see https://www.opensips.org/Documentation/TroubleShooting-OutOfMem, the "How to handle it" section.
After recompiling and running OpenSIPS, we only need to wait for another crash.

**Comment by netaskd** [comment]
version:
```
opensips -V
version: opensips 2.4.5 (x86_64/linux)
flags: STATS: On, DISABLE_NAGLE, USE_MCAST, SHM_MMAP, PKG_MALLOC, QM_MALLOC, DBG_MALLOC, FAST_LOCK-ADAPTIVE_WAIT
ADAPTIVE_WAIT_LOOPS=1024, MAX_RECV_BUFFER_SIZE 262144, MAX_LISTEN 16, MAX_URI_SIZE 1024, BUF_SIZE 65535
poll method support: poll, epoll, sigio_rt, select.
main.c compiled on 14:20:17 Mar 26 2019 with gcc 7
```
log:
```
Mar 26 14:57:44 [760] CRITICAL:core:qm_free: bad pointer 0x7f2e0a937808 (out of memory block!) - aborting
Mar 26 14:57:45 [770] CRITICAL:core:handle_worker: dead child 9 (EOF received), pid 760
Mar 26 14:57:45 [744] INFO:core:handle_sigs: child process 760 exited by a signal 6
Mar 26 14:57:45 [744] INFO:core:handle_sigs: core was generated
Mar 26 14:57:45 [744] INFO:core:handle_sigs: terminating due to SIGCHLD
Mar 26 14:57:45 [754] INFO:core:sig_usr: signal 15 received
```
core: https://pastebin.com/Wtx6KTa3

**Comment by netaskd** [comment]
f 5 p *p_cell https://pastebin.com/7BTpRmkn

**Comment by bogdan-iancu** [comment]
Thanks @netaskd  - please print in frame 4 `p *dead_cell->uas.request`

**Comment by netaskd** [comment]
f 4 p *dead_cell->uas.request:  https://pastebin.com/CCg2rWaN

**Comment by bogdan-iancu** [comment]
@netaskd , I guess we re-discovered an existing issue - the "path_vec" member of the SIP UAS request inside the transaction points to the wrong memory zone (the whole request is in shm memeory, but the `path_vec` is in the private memory. This is a duplicate of #1624 (and I will close that one).

I understand you proxy REGISTER requests and you use PATH support, right ?

**Comment by netaskd** [comment]
@bogdan-iancu thanks for the reply.
I use proxy, including, as a b2b registrar. In first I try to authenticate locally, and if its passed, dont send 200 onreply for now, but instead build a new auth request and send it to a backend registrar (not opensips) for auth request, and only after the backend return 200, i send 200 to uac from proxy
will try explain via showing logic that looks like:
```
loadmodule "registrar.so"
...
modparam("registrar", "tcp_persistent_flag", "sip_tcp")
modparam("registrar", "received_avp", "$avp(received)")
modparam("registrar", "attr_avp", "$avp(attr)")
modparam("registrar", "gruu_secret", "XXX")
modparam("registrar", "case_sensitive", 1)
modparam("registrar", "default_q", 380)

loadmodule "path.so"
modparam("path", "use_received", 1)
modparam("path", "enable_double_path", 1)

#### User Agent Client module #######################
loadmodule "uac_auth.so"
loadmodule "uac.so"
modparam("uac","restore_mode","none")
modparam("uac_auth","auth_realm_avp","$avp(uas_realm)")
modparam("uac_auth","auth_username_avp","$avp(uas_username)")
modparam("uac_auth","auth_password_avp","$avp(uas_password)")

route["INITIAL_REQUESTS"] {
...
  if (is_method("REGISTER")) {
  ...
  route("REGISTRAR");
  }
...
}

route[REGISTRAR] {
  if (!add_path_received()) {
    xlog("L_ERR", "[REGISTRAR] $rm - $ci unable to add path. stop processing\n");
    sl_send_reply("503", "Internal path befuddlement");
    exit;
  }

  ...
 # auth locally, but not send 200 when it's passed
 route("AUTH_LOC");
  ....
 
  # save To as aor for register only
  # catch 200 register from remote uas
  # and save location here
  t_on_reply("ONREPLY_B2BUA_SAVE_AOR");

  # b-leg uac auth here
  t_on_failure("B2BUA_AUTH_GEN");

  route("RELAY");
}

route[AUTH_LOC] {
  # use cache fro pw
    if(cache_fetch("local", "auth_pw_$fU", $avp(auth_pw))) {
      xlog("L_INFO", "[AUTH_OR_BLOCK] $rm - $C(gx)got creds for auth_pw_$(fU) from cache$C(xx)\n");
      $avp(auth_un) = $fU;
      $var(rc) = pv_www_authorize("");
    } else {
      $var(rc) = www_authorize("","ro_users");
    }
    switch ($var(rc)) {
     case -4:  # no credentials
      case -2:  # invalid passwd
      if(is_method("REGISTER")) {
        www_challenge("", "0");
      } else {
        proxy_challenge("", "0");
      }
   }; # user authenticated
  cache_store("local","auth_pw_$fU","$avp(auth_pw)"); # save to cache
  consume_credentials(); # do not send credential to next hop
  ...
}

onreply_route[ONREPLY_B2BUA_SAVE_AOR] {
 if (t_check_status("200")) {
    if(is_method("REGISTER")) {
      $var(sflg) = "p0c1vfr"; # save path lazy mode (p1), max_contact is 1 (c1), path from receiVed param (v), force update when new register (f), no reply (r)
      construct_uri("sip", "$tU", "$td", "$avp(ct:uri_port)", "transport=$avp(ct:uri_proto)",$avp(aor));
      if (save("location","$var(sflg)","$avp(aor)")) {
         # will send 200 to UAC
        ...
      }
    }
 } 
}

failure_route[B2BUA_AUTH_GEN] { # re-routing for 40[17] auth requests

  # authentication reply received?
  if (!t_check_status("40[17]")) { # only for 401/407 auth requests
     ...
    return;
  }

  # have we already tried to authenticate?
  if (isflagset(WAS_AUTHED)) {
   ... 
   drop;
  }
  ...

  # mark that auth was performed
  setflag(WAS_AUTHED);
  # arm again the failure route
  t_on_failure("B2BUA_AUTH_GEN");

  # get Authenticate header from reply context
  if ($(<reply>hdr(Proxy-Authenticate))) {
    $var(raw_auth) = $(<reply>hdr(Proxy-Authenticate));
  } else if ($(<reply>hdr(WWW-Authenticate))) {
    $var(raw_auth) = $(<reply>hdr(WWW-Authenticate));
  }
  
  if ($var(raw_auth)) {
    xlog("L_INFO", "Raw Authenticate data '$var(raw_auth)'. Try to get uac realm...\n")
    # some regexp operation
     ....
    xlog("L_INFO", "Got reply uas_realm=$avp(uas_realm), request realm=$ar\n");
     ...
  }
  ...
  # here we have $avp(uas_realm)
  $avp(uas_username) = $fU;
  $avp(uas_password) = $sql_cached_value(uas:password:$avp(uas_username));
  if ($avp(uas_password)) {
     if (uac_auth()) { # build authentication and send it to backend UAS
       $du = $avp(backend_dst);
       route("RELAY");
       exit; 
     }
  }
}
```

**Comment by bogdan-iancu** [comment]
Thanks @netaskd  - is it too much if I'm asking for your cfg file (off ticket, with full privacy, my eyes only) ? I want to see the transaction related ops that may be done while the REGISTER is handled.

**Comment by netaskd** [comment]
@bogdan-iancu I sent to your email.

**Comment by bogdan-iancu** [comment]
Many thanks @netaskd for providing all the needed data for finding this issue !
